src/main/scala/com/exasol/cloudetl/kinesis/KinesisShardDataImporter.scala

```diff
@@ -0,0 +1,116 @@
+      awsAccessKeyId,
+      awsSecretAccessKey,
+      awsSessionToken,
+      streamName,
+      region,
+      shardId,
+      shardSequenceNumber
+    )
+    try {
+      records.asScala.foreach(
+        record => {
+          val values = createTableValuesListFromRecord(record, shardId)
+          val columns: Seq[Object] = values.map(_.asInstanceOf[AnyRef])
+          exaIterator.emit(columns: _*)
+        }
+      )
+    } catch {
+      case exception @ (_: ExaDataTypeException | _: ExaIterationException) =>
+        throw new KinesisConnectorException(
+          "KinesisShardDataImporter cannot emit records. Caused: " + exception.getMessage,
+          exception
+        )
+    }
+  }
+
+  private def importRecordsForShard(
+    awsAccessKeyId: String,
+    awsSecretAccessKey: String,
+    awsSessionToken: String,
+    streamName: String,
+    region: String,
+    shardId: String,
+    shardSequenceNumber: String
+  ): util.List[Record] = {
+    val awsCredentials = getAwsCredentials(awsAccessKeyId, awsSecretAccessKey, awsSessionToken)
+    val amazonKinesis = AmazonKinesisClientBuilder.standard
+      .withRegion(region)
+      .withCredentials(new AWSStaticCredentialsProvider(awsCredentials))
+      .build
+    val shardIteratorRequest =
+      createShardIteratorRequest(shardId, shardSequenceNumber, streamName)
+    val shardIteratorResult = amazonKinesis.getShardIterator(shardIteratorRequest)
+    val shardIterator = shardIteratorResult.getShardIterator
+    getRecords(amazonKinesis, shardIterator)
+  }
+
+  private def getAwsCredentials(
+    awsAccessKeyId: String,
+    awsSecretAccessKey: String,
+    awsSessionToken: String
+  ) = new BasicSessionCredentials(awsAccessKeyId, awsSecretAccessKey, awsSessionToken)
+
+  private def createShardIteratorRequest(
+    shardId: String,
+    shardSequenceNumber: String,
+    streamName: String
+  ): GetShardIteratorRequest = {
+    val getShardIteratorRequest = new GetShardIteratorRequest
+    getShardIteratorRequest.setStreamName(streamName)
+    getShardIteratorRequest.setShardId(shardId)
+    if (shardSequenceNumber != null) {
+      getShardIteratorRequest.setShardIteratorType(ShardIteratorType.AFTER_SEQUENCE_NUMBER)
+      getShardIteratorRequest.setStartingSequenceNumber(shardSequenceNumber)
+    } else {
+      getShardIteratorRequest.setShardIteratorType(ShardIteratorType.TRIM_HORIZON)
+    }
+    getShardIteratorRequest
+  }
+
+  private def getRecords(
+    amazonKinesis: AmazonKinesis,
+    shardIterator: String
+  ): util.List[Record] = {
+    val getRecordsRequest = new GetRecordsRequest
+    getRecordsRequest.setShardIterator(shardIterator)
+    val result = amazonKinesis.getRecords(getRecordsRequest)
+    result.getRecords
+  }
+
+  @SuppressWarnings(Array("org.wartremover.warts.NonUnitStatements"))
+  def createTableValuesListFromRecord(record: Record, shardId: String): Seq[Any] = {
+    val data = record.getData
+    val mapper = new ObjectMapper with ScalaObjectMapper
```

Comment:
Why do we name it `fromJson`? Should I also move parsing logic or is it only returns a mapper?

Reply:
Yes. The idea is to avoid creating mapper for each record. Using the static object class you create it once, and call the parsing method multiple times. Please feel free to use clear naming for the method, `parseJsonString`, `parseFromJson` or other. It is important that the function is parameterized with a type.

```scala
object JsonUtils {
lazy val mapper = new ObjectMapper with ScalaObjectMapper
// mapper initialization

def fromJson[T](jsonString: String): T = {
mapper.readValue[T](jsonString)
}
}
```